charts/posit-chronicle/templates/stateful-set.yaml

```diff
@@ -0,0 +1,80 @@
+---
+apiVersion: apps/v1
+kind: StatefulSet
```

Comment:
It's mostly as a convenience for the pvc provisioning, if you're running with local storage enabled Persistent storage a pvc is a must-have for most use cases. It's nice to not have to worry about duplicate replica sets being created on helm updates as well, the pod replacement policy of stateful-sets is generally simple and non-problematic for chronicle.

Theoretically we could run chronicle on a deployment/replica-set model, there's not too many technical blockers to think of. The stateful-set model has worked well so far though, and our local storage situation will be a bit trickier without it.

Reply:
That makes sense.

It's worth noting that statefulsets do have some downsides: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#limitations

Usually services that uses statefulsets have a particular reason to do so. Otherwise replicasets/deployments are the more standard choice.

If you use a PVC, does each chronicle node store its own data? Or do they share data storage?

Reply:
That's tough, we do have a choice to make for sure. As a developer I feel like it's nice to have those pvcs deleted by default:

> Deleting and/or scaling a StatefulSet down will not delete the volumes associated with the StatefulSet. This is done to ensure data safety, which is generally more valuable than an automatic purge of all related StatefulSet resources.

Though on the other hand, we are trying to build a data archive too. If we scaled a chronicle server from 1->5 replicas for a day, and stored a bunch of data, it would be nice if 2 weeks later when we tried scaling 1->5 again we found our old data in those local storage systems that we could add to. 

> If you use a PVC, does each chronicle node store its own data? Or do they share data storage?

Yeah each chronicle server does need a dedicated pvc, that's a known technical limitation right now in the alpha though. We need to rewrite some in-memory locks into file-locks for our compactor engine, and then we can support nfs or shared storage. We'll definitely need to make some helm changes around that too.